app/src/main/java/org/mozilla/fenix/utils/VisualMetricsInstrumentation.kt

```diff
@@ -0,0 +1,96 @@
+package org.mozilla.fenix.utils
+
+import android.app.Instrumentation
+import android.graphics.Color
+import android.graphics.drawable.ColorDrawable
+import android.os.Build
+import android.os.Bundle
+import android.util.Log
+import android.view.KeyEvent
+import android.view.ViewGroup
+import androidx.annotation.RequiresApi
+import androidx.core.view.children
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentManager
+import kotlinx.android.synthetic.main.fragment_home.view.*
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.runBlocking
+import mozilla.components.browser.toolbar.BrowserToolbar
+import mozilla.components.ui.autocomplete.InlineAutocompleteEditText
+import org.mozilla.fenix.R
+import org.mozilla.fenix.browser.BaseBrowserFragment
+import org.mozilla.fenix.browser.BrowserFragment
+import org.mozilla.fenix.home.HomeFragment
+import org.mozilla.fenix.search.SearchFragment
+import java.util.concurrent.TimeUnit
+
+@ExperimentalCoroutinesApi
+@RequiresApi(Build.VERSION_CODES.M)
+class VisualMetricsInstrumentation(private val naviageToUrl: String) :
+    FragmentManager.FragmentLifecycleCallbacks() {
+
+    private var runnableEvent: Thread
+
+    init {
+        runnableEvent = Thread {
+            try {
+                val inst = Instrumentation()
+                inst.sendKeyDownUpSync(KeyEvent.KEYCODE_ENTER)
+            } catch (e: InterruptedException) {
+                Log.e("error dispatching key", "test failed")
+            }
+        }
+    }
+
+    override fun onFragmentCreated(fm: FragmentManager, f: Fragment, savedInstanceState: Bundle?) {
+        super.onFragmentCreated(fm, f, savedInstanceState)
+        if (f is SearchFragment) {
+            runBlocking {
+                f.activity!!.window.decorView.foreground = ColorDrawable(Color.GREEN)
+            }
+        } else if (f is BaseBrowserFragment) {
+            runBlocking {
+                f.activity!!.window.decorView.foreground = ColorDrawable(Color.RED)
+            }
+        }
+    }
+
+    override fun onFragmentResumed(fm: FragmentManager, f: Fragment) {
+        super.onFragmentResumed(fm, f)
+        if (f is HomeFragment) {
+            f.view!!.post {
+                f.activity!!.window.decorView.foreground = null
+                f.view!!.postOnAnimation {
+                    f.view!!.toolbar_wrapper.callOnClick()
+                }
+            }
+        } else if (f is SearchFragment) {
```

Comment:
Why are we going through this workflow?

In general, I don't love shipping code not intended for users with the APK – it bloats the code and creates a security risk – and fake pressing UI elements is throwing it over the line for me. Is there another way to do this? e.g. with an espresso test?

Reply:
As much as I agree, using espresso test 1) doesn't give real performance metrics that a user would see. 2) We shouldn't have this code in,  100% agree. 3) This is 100% not an idea solution. It should be probably build dependent where we don't ship this code in real production (i.e have one build that is _exactly_ like prod but only contains instrumentation).

Imo the final solution to this would be using [Monkey](https://developer.android.com/studio/test/monkeyrunner) since we don't need to interact with GeckoDriver but only need to instrument user interactions. I wouldn't be for using Espresso apk to test and record. That would lead to a lot of overhead that doesn't represent  our real flow

Reply:
> using espresso test 1) doesn't give real performance metrics that a user would see.
>
> I wouldn't be for using Espresso apk to test and record. That would lead to a lot of overhead that doesn't represent our real flow

I'm not familiar with Espresso's overhead – where does that come from?

> 3) This is 100% not an idea solution. 

We should probably document this (in code comments?) – and then we wouldn't be having this conversation. 😜